statsmodels/tsa/statespace/regime_switching/kim_filter.py

```diff
@@ -0,0 +1,635 @@
+import numpy as np
+from scipy.misc import logsumexp
+from scipy.optimize import linprog
+from statsmodels.tsa.statespace.kalman_filter import KalmanFilter
+
+try:
+    from scipy.stats import multivariate_normal
+    multivariate_normal_logpdf = multivariate_normal.logpdf
+except ImportError:
+    def multivariate_normal_logpdf(x, mean=None, cov=None):
+        n = x.shape[0]
+        if np.linalg.matrix_rank(cov) < n:
+            raise RuntimeError('Cov matrix is singular.')
+        #TODO: to test this
+        return -0.5 * n * np.log(np.pi) - 0.5 * np.log(np.linalg.det(cov)) - \
+                0.5 * (x - mean).T.dot(np.linalg.inv(cov)).dot(x - mean)
+
+class KimFilter(object):
+    '''
+    Kim Filter
+    '''
+
+    def __init__(self, k_endog, k_states, k_regimes, dtype=np.float64,
+            loglikelihood_burn=0, designs=None, obs_intercepts=None,
+            obs_covs=None, transitions=None, state_intercepts=None,
+            selections=None, state_covs=None, regime_switch_probs=None,
+            **kwargs):
+
+        if k_regimes < 1:
+            raise ValueError('Only multiple regimes are available.'
+                    'Consider using regular KalmanFilter.')
+
+        self._k_endog = k_endog
+        self._k_states = k_states
+        self._k_regimes = k_regimes
+        self._dtype = dtype
```

Comment:
`dtype` can change between `filter` calls, e.g. when we do complex-step differentiation and the parameter vector becomes complex. Do you create a new KimFilter object each time the parameter vector changes?